### PR TITLE
feat: dataspace protocol negotiation transformer

### LINKED ISSUE #2764: Dataspace-Protocol: Missing attributes code and reason in TransferTerminationMessage and ContractNegotiationTerminationMessage
# Feature Request

The request message `TransferTerminationMessage` misses the attributes code and reason like the [spec](https://docs.internationaldataspaces.org/dataspace-protocol/transfer-process/transfer.process.protocol#5.-transferterminationmessage) has.

The request message `ContractNegotiationTerminationMessage` misses the attribute code like the [spec](https://docs.internationaldataspaces.org/dataspace-protocol/contract-negotiation/contract.negotiation.protocol#6.-contractnegotiationterminationmessage) defines.

## Which Areas Would Be Affected?
control-plane, dsp-specification

## Why Is the Feature Desired?
support the dataspace protocol

## Solution Proposal
Add the fields `code` and `reason` to the `TransferTerminationMessage`.

Add the field `code` to the `TransferTerminationMessage`.

Also we have to look on the discussion in #2729 and look how the reason attribute should look like in the future.

### PR DESCRIPTION
**Pull request by juliapampus**
## What this PR changes/adds

Implements contract negotiation spi and transformers:
- `:data-protocols:dsp:dsp-negotiation:dsp-negotiation-spi`
- `:data-protocols:dsp:dsp-negotiation:dsp-negotiation-transform`

spec alignments:
- remove null check for `callback` in `ContractAgreementMessage`
- add checksum to `ContractNegotiationEventMessage`
- add `code` in `ContractNegotiationTerminationMessage`
- remove null check for `rejectionReason` in `ContractNegotiationTerminationMessage`
- deprecate `contractStart` and `contractEnd` in `ContractOffer`, see #2758
- add `checksum` to `ContractNegotiation`

## Why it does that

Support the dataspace protocol specification.

## Further notes

Derived issue #2775.

## Linked Issue(s)

Relates #2474
Closes #2764

## Checklist

- [x] added appropriate tests?
- [x] performed checkstyle check locally?
- [x] added/updated copyright headers?
- [x] documented public classes/methods?
- [ ] added/updated relevant documentation?
- [x] assigned appropriate label? (exclude from changelog with label `no-changelog`)
- [x] formatted title correctly? (_take a look at the [CONTRIBUTING](https://github.com/eclipse-edc/Connector/blob/main/CONTRIBUTING.md#submit-a-pull-request) and [Etiquette for pull requests](https://github.com/eclipse-edc/Connector/blob/main/pr_etiquette.md) for details_)

### CODE REVIEW

**Comment by juliapampus** [comment]
Open question: How to set `providerAgentId` and `consumerAgentId` in `ContractAgreement`? @jimmarino

File: data-protocols/dsp/dsp-negotiation/dsp-negotiation-transform/src/main/java/org.eclipse.edc.protocol.dsp.negotiation.transform/DspNegotiationTransformExtension.java [hunk 1]
```diff
@@ -0,0 +1,74 @@
...
+import org.eclipse.edc.runtime.metamodel.annotation.Extension;
+import org.eclipse.edc.runtime.metamodel.annotation.Inject;
+import org.eclipse.edc.runtime.metamodel.annotation.Provides;
+import org.eclipse.edc.spi.system.ServiceExtension;
+import org.eclipse.edc.spi.system.ServiceExtensionContext;
+import org.eclipse.edc.spi.types.TypeManager;
+
+import java.util.Map;
+
+@Extension(value = JsonLdExtension.NAME)
```
Comment (ronjaquensel) [inline data-protocols/dsp/dsp-negotiation/dsp-negotiation-transform/src/main/java/org.eclipse.edc.protocol.dsp.negotiation.transform/DspNegotiationTransformExtension.java:40]: should be `DspNegotiationTransformExtension.NAME`

File: data-protocols/dsp/dsp-negotiation/dsp-negotiation-transform/src/main/java/org.eclipse.edc.protocol.dsp.negotiation.transform/DspNegotiationTransformExtension.java [hunk 2]
```diff
@@ -0,0 +1,74 @@
...
+@Provides({JsonLdTransformerRegistry.class})
```
Comment (ronjaquensel) [inline data-protocols/dsp/dsp-negotiation/dsp-negotiation-transform/src/main/java/org.eclipse.edc.protocol.dsp.negotiation.transform/DspNegotiationTransformExtension.java:41]: The `@Provides` should be removed

File: data-protocols/dsp/dsp-negotiation/dsp-negotiation-transform/src/main/java/org.eclipse.edc.protocol.dsp.negotiation.transform/from/JsonObjectFromContractAgreementMessageTransformer.java [hunk 3]
```diff
@@ -0,0 +1,68 @@
...
+            context.reportProblem("Cannot transform from ContractAgreementMessage with null policy");
+            return null;
+        }
+
+        builder.add(DSPACE_NEGOTIATION_PROPERTY_AGREEMENT, policy);
+
+        return builder.build();
+    }
+
+    private @Nullable JsonObject transformPolicy(ContractAgreement agreement, TransformerContext context) {
```
Comment (ndr-brt) [inline data-protocols/dsp/dsp-negotiation/dsp-negotiation-transform/src/main/java/org.eclipse.edc.protocol.dsp.negotiation.transform/from/JsonObjectFromContractAgreementMessageTransformer.java:64]: inline

File: data-protocols/dsp/dsp-negotiation/dsp-negotiation-transform/src/main/java/org.eclipse.edc.protocol.dsp.negotiation.transform/from/JsonObjectFromContractRequestTransformer.java [hunk 4]
```diff
@@ -0,0 +1,72 @@
...
+            context.reportProblem("Cannot transform from ContractRequestMessage with null policy");
+            return null;
+        }
+
+        builder.add(DSPACE_NEGOTIATION_PROPERTY_OFFER, policy);
+
+        return builder.build();
+    }
+
+    private @Nullable JsonObject transformPolicy(ContractOffer offer, TransformerContext context) {
```
Comment (ndr-brt) [inline data-protocols/dsp/dsp-negotiation/dsp-negotiation-transform/src/main/java/org.eclipse.edc.protocol.dsp.negotiation.transform/from/JsonObjectFromContractRequestTransformer.java:68]: inline

File: data-protocols/dsp/dsp-negotiation/dsp-negotiation-transform/src/main/java/org.eclipse.edc.protocol.dsp.negotiation.transform/to/JsonObjectToContractAgreementMessageTransformer.java [hunk 5]
```diff
@@ -0,0 +1,85 @@
...
+            context.reportProblem("Cannot transform to ContractAgreementMessage with null agreement");
+            return null;
+        }
+
+        builder.contractAgreement(agreement);
+
+        return builder.build();
+    }
+
+    private @Nullable Policy transformPolicy(JsonObject policy, TransformerContext context) {
```
Comment (ndr-brt) [inline data-protocols/dsp/dsp-negotiation/dsp-negotiation-transform/src/main/java/org.eclipse.edc.protocol.dsp.negotiation.transform/to/JsonObjectToContractAgreementMessageTransformer.java:63]: inline

File: data-protocols/dsp/dsp-negotiation/dsp-negotiation-transform/src/main/java/org.eclipse.edc.protocol.dsp.negotiation.transform/to/JsonObjectToContractNegotiationEventMessageTransformer.java [hunk 6]
```diff
@@ -0,0 +1,78 @@
...
+import static org.eclipse.edc.protocol.dsp.negotiation.spi.DspNegotiationPropertyAndTypeNames.DSPACE_NEGOTIATION_PROPERTY_PROCESS_ID;
+import static org.eclipse.edc.protocol.dsp.spi.types.HttpMessageProtocol.DATASPACE_PROTOCOL_HTTP;
+
+/**
+ * Creates a {@link ContractNegotiationEventMessage} from a {@link JsonObject}.
+ */
+public class JsonObjectToContractNegotiationEventMessageTransformer extends AbstractJsonLdTransformer<JsonObject, ContractNegotiationEventMessage> {
+
+    private TransformerContext context;
+    private ContractNegotiationEventMessage.Type eventType;
```
Comment (ndr-brt) [inline data-protocols/dsp/dsp-negotiation/dsp-negotiation-transform/src/main/java/org.eclipse.edc.protocol.dsp.negotiation.transform/to/JsonObjectToContractNegotiationEventMessageTransformer.java:39]: this should be a method-scoped variable, not a field

File: data-protocols/dsp/dsp-negotiation/dsp-negotiation-transform/build.gradle.kts [hunk 7]
```diff
@@ -0,0 +1,26 @@
...
+ */
+
+plugins {
+    `java-library`
+}
+
+dependencies {
+    api(project(":data-protocols:dsp:dsp-http-spi"))
+    api(project(":data-protocols:dsp:dsp-negotiation:dsp-negotiation-spi"))
+    api(project(":data-protocols:dsp:dsp-transform"))
```
Comment (ndr-brt) [inline data-protocols/dsp/dsp-negotiation/dsp-negotiation-transform/build.gradle.kts:22]: this dependency is suspect, I think we'd need a `dsp-transform-spi` module, maybe something for subsequent PRs

**Review by ndr-brt — COMMENTED** [review]
pretty much ok, apart from that potential bug in `JsonObjectToContractNegotiationEventMessageTransformer`, we should remove the side effect from `transformString` and make it return the string instead of evaluating a function passed.
Maybe first adding a new method, to avoid bloating this PR.

File: data-protocols/dsp/dsp-negotiation/dsp-negotiation-spi/build.gradle.kts [hunk 8]
```diff
@@ -0,0 +1,21 @@
...
+ *       Fraunhofer Institute for Software and Systems Engineering - initial API and implementation
+ *
+ */
+
+plugins {
+    `java-library`
+}
+
+dependencies {
+    api(project(":spi:common:core-spi"))
```
Comment (jimmarino) [inline data-protocols/dsp/dsp-negotiation/dsp-negotiation-spi/build.gradle.kts:20]: does this module need to depend on core spi?

File: data-protocols/dsp/dsp-negotiation/dsp-negotiation-transform/src/main/java/org.eclipse.edc.protocol.dsp.negotiation.transform/from/JsonObjectFromContractAgreementVerificationMessageTransformer.java [hunk 9]
```diff
@@ -0,0 +1,54 @@
...
+    }
+
+    @Override
+    public @Nullable JsonObject transform(@NotNull ContractAgreementVerificationMessage object, @NotNull TransformerContext context) {
+        var builder = jsonFactory.createObjectBuilder();
+        builder.add(JsonLdKeywords.ID, String.valueOf(UUID.randomUUID()));
+        builder.add(JsonLdKeywords.TYPE, DSPACE_NEGOTIATION_AGREEMENT_VERIFICATION_MESSAGE);
+
+        builder.add(DSPACE_NEGOTIATION_PROPERTY_PROCESS_ID, object.getProcessId());
+        // TODO add mapping of cred:credentialSubject and sec:proof for signature processes
```
Comment (jimmarino) [inline data-protocols/dsp/dsp-negotiation/dsp-negotiation-transform/src/main/java/org.eclipse.edc.protocol.dsp.negotiation.transform/from/JsonObjectFromContractAgreementVerificationMessageTransformer.java:50]: The proofs aren't supposed to be in the spec (they were examples) so I would remove that from the TODO

File: data-protocols/dsp/dsp-negotiation/dsp-negotiation-transform/src/main/java/org.eclipse.edc.protocol.dsp.negotiation.transform/from/JsonObjectFromContractNegotiationEventMessageTransformer.java [hunk 10]
```diff
@@ -0,0 +1,79 @@
...
+    }
+
+    private String eventType(ContractNegotiationEventMessage message, TransformerContext context) {
+        switch (message.getType()) {
+            case ACCEPTED:
+                return DSPACE_NEGOTIATION_PROPERTY_EVENT_TYPE_ACCEPTED;
+            case FINALIZED:
+                return DSPACE_NEGOTIATION_PROPERTY_EVENT_TYPE_FINALIZED;
+            default:
+                context.reportProblem(String.format("Could not map eventType %s", message.getType()));
```
Comment (jimmarino) [inline data-protocols/dsp/dsp-negotiation/dsp-negotiation-transform/src/main/java/org.eclipse.edc.protocol.dsp.negotiation.transform/from/JsonObjectFromContractNegotiationEventMessageTransformer.java:74]: I would be good to have more context, e.g. in contract negotiation event message

File: data-protocols/dsp/dsp-negotiation/dsp-negotiation-transform/src/main/java/org.eclipse.edc.protocol.dsp.negotiation.transform/to/JsonObjectToContractNegotiationEventMessageTransformer.java [hunk 11]
```diff
@@ -0,0 +1,78 @@
...
+    @Override
+    public @Nullable ContractNegotiationEventMessage transform(@NotNull JsonObject object, @NotNull TransformerContext context) {
+        this.context = context;
+
+        var builder = ContractNegotiationEventMessage.Builder.newInstance();
+        builder.protocol(DATASPACE_PROTOCOL_HTTP);
+        transformString(object.get(DSPACE_NEGOTIATION_PROPERTY_PROCESS_ID), builder::processId, context);
+        transformString(object.get(DSPACE_NEGOTIATION_PROPERTY_CHECKSUM), builder::checksum, context);
+
+        transformString(object.get(DSPACE_NEGOTIATION_PROPERTY_EVENT_TYPE), this::type, context);
```
[the remark this answers is not in the document]
Comment (jimmarino) [reply data-protocols/dsp/dsp-negotiation/dsp-negotiation-transform/src/main/java/org.eclipse.edc.protocol.dsp.negotiation.transform/to/JsonObjectToContractNegotiationEventMessageTransformer.java:54]: The one issue with this is in the case that a transformation can't be made, null would have to be returned and all invocations would have to explicitly handle that case

File: data-protocols/dsp/dsp-negotiation/dsp-negotiation-transform/src/main/java/org.eclipse.edc.protocol.dsp.negotiation.transform/to/JsonObjectToContractNegotiationTerminationMessageTransformer.java [hunk 12]
```diff
@@ -0,0 +1,58 @@
...
+        transformString(object.get(DSPACE_NEGOTIATION_PROPERTY_PROCESS_ID), builder::processId, context);
+        transformString(object.get(DSPACE_NEGOTIATION_PROPERTY_CODE), builder::code, context);
+
+        var reasons = object.get(DSPACE_NEGOTIATION_PROPERTY_REASON);
+        if (reasons != null) {
+            var result = typeValueArray(reasons, context);
+            if (result == null) {
+                context.reportProblem(format("Cannot transform property %s", DSPACE_NEGOTIATION_PROPERTY_REASON));
+            } else {
+                builder.rejectionReason(String.valueOf(result.get(0)));
```
Comment (jimmarino) [inline data-protocols/dsp/dsp-negotiation/dsp-negotiation-transform/src/main/java/org.eclipse.edc.protocol.dsp.negotiation.transform/to/JsonObjectToContractNegotiationTerminationMessageTransformer.java:52]: should check index bounds

**Review by jimmarino — APPROVED** [review]
Just some specific changes, otherwise for me OK pending others' comments.

File: data-protocols/dsp/dsp-negotiation/dsp-negotiation-transform/src/main/java/org.eclipse.edc.protocol.dsp.negotiation.transform/from/JsonObjectFromContractRequestTransformer.java (hunk 4 above)
Comment (janpmeyer) [reply data-protocols/dsp/dsp-negotiation/dsp-negotiation-transform/src/main/java/org.eclipse.edc.protocol.dsp.negotiation.transform/from/JsonObjectFromContractRequestTransformer.java]: done

File: data-protocols/dsp/dsp-negotiation/dsp-negotiation-transform/src/main/java/org.eclipse.edc.protocol.dsp.negotiation.transform/to/JsonObjectToContractAgreementMessageTransformer.java (hunk 5 above)
Comment (janpmeyer) [reply data-protocols/dsp/dsp-negotiation/dsp-negotiation-transform/src/main/java/org.eclipse.edc.protocol.dsp.negotiation.transform/to/JsonObjectToContractAgreementMessageTransformer.java]: done

File: data-protocols/dsp/dsp-negotiation/dsp-negotiation-transform/src/main/java/org.eclipse.edc.protocol.dsp.negotiation.transform/to/JsonObjectToContractNegotiationEventMessageTransformer.java (hunk 6 above)
Comment (janpmeyer) [reply data-protocols/dsp/dsp-negotiation/dsp-negotiation-transform/src/main/java/org.eclipse.edc.protocol.dsp.negotiation.transform/to/JsonObjectToContractNegotiationEventMessageTransformer.java]: Field is removed

File: data-protocols/dsp/dsp-negotiation/dsp-negotiation-transform/src/main/java/org.eclipse.edc.protocol.dsp.negotiation.transform/to/JsonObjectToContractNegotiationEventMessageTransformer.java (hunk 11 above)
[the remark this answers is not in the document]
Comment (janpmeyer) [reply data-protocols/dsp/dsp-negotiation/dsp-negotiation-transform/src/main/java/org.eclipse.edc.protocol.dsp.negotiation.transform/to/JsonObjectToContractNegotiationEventMessageTransformer.java]: type function is wrapped up in a lambda expression. Hope this way is fine for you. If not i adapt your comments.

File: data-protocols/dsp/dsp-negotiation/dsp-negotiation-transform/src/main/java/org.eclipse.edc.protocol.dsp.negotiation.transform/to/JsonObjectToContractRequestMessageTransformer.java [hunk 13]
```diff
@@ -0,0 +1,76 @@
...
+            context.reportProblem("Cannot transform to ContractRequestMessage with null policy");
+            return null;
+        }
+
+        builder.contractOffer(contractOffer(object, policy));
+
+        return builder.build();
+    }
+
+    private @Nullable Policy transformPolicy(JsonObject policy, TransformerContext context) {
```
[the remark this answers is not in the document]
Comment (janpmeyer) [reply data-protocols/dsp/dsp-negotiation/dsp-negotiation-transform/src/main/java/org.eclipse.edc.protocol.dsp.negotiation.transform/to/JsonObjectToContractRequestMessageTransformer.java]: done

File: data-protocols/dsp/dsp-negotiation/dsp-negotiation-transform/build.gradle.kts (hunk 7 above)
Comment (janpmeyer) [reply data-protocols/dsp/dsp-negotiation/dsp-negotiation-transform/build.gradle.kts:22]: Noted for a later PR.

File: data-protocols/dsp/dsp-negotiation/dsp-negotiation-spi/build.gradle.kts (hunk 8 above)
Comment (janpmeyer) [reply data-protocols/dsp/dsp-negotiation/dsp-negotiation-spi/build.gradle.kts]: Removed dependency

File: data-protocols/dsp/dsp-negotiation/dsp-negotiation-transform/src/main/java/org.eclipse.edc.protocol.dsp.negotiation.transform/from/JsonObjectFromContractAgreementVerificationMessageTransformer.java (hunk 9 above)
Comment (janpmeyer) [reply data-protocols/dsp/dsp-negotiation/dsp-negotiation-transform/src/main/java/org.eclipse.edc.protocol.dsp.negotiation.transform/from/JsonObjectFromContractAgreementVerificationMessageTransformer.java]: Removed "and sec:proof from TODO"

File: data-protocols/dsp/dsp-negotiation/dsp-negotiation-transform/src/main/java/org.eclipse.edc.protocol.dsp.negotiation.transform/from/JsonObjectFromContractNegotiationEventMessageTransformer.java (hunk 10 above)
Comment (janpmeyer) [reply data-protocols/dsp/dsp-negotiation/dsp-negotiation-transform/src/main/java/org.eclipse.edc.protocol.dsp.negotiation.transform/from/JsonObjectFromContractNegotiationEventMessageTransformer.java]: Extended to "Could not map eventType %s to %s or %s in ContractNegotiationEventMessage"

File: data-protocols/dsp/dsp-negotiation/dsp-negotiation-transform/src/main/java/org.eclipse.edc.protocol.dsp.negotiation.transform/to/JsonObjectToContractNegotiationTerminationMessageTransformer.java (hunk 12 above)
Comment (janpmeyer) [reply data-protocols/dsp/dsp-negotiation/dsp-negotiation-transform/src/main/java/org.eclipse.edc.protocol.dsp.negotiation.transform/to/JsonObjectToContractNegotiationTerminationMessageTransformer.java]: Checking if size of Array is bigger 0.

File: data-protocols/dsp/dsp-negotiation/dsp-negotiation-transform/build.gradle.kts (hunk 7 above)
Comment (juliapampus) [reply data-protocols/dsp/dsp-negotiation/dsp-negotiation-transform/build.gradle.kts]: dependency is not needed. Removed it. Still, this is a valid suggestion, created #2805.

File: data-protocols/dsp/dsp-negotiation/dsp-negotiation-spi/build.gradle.kts [hunk 14]
```diff
@@ -17,5 +17,4 @@ plugins {
 }
 
 dependencies {
```
Comment (ndr-brt) [inline data-protocols/dsp/dsp-negotiation/dsp-negotiation-spi/build.gradle.kts:19]: the whole `dependencies` block can be removed

**Review by ndr-brt — APPROVED** [review]

**Pull request merged**